File: Views/ArticleListViewAdapter.cs
using System;
using System.IO;
using System.Threading;
using System.Collections.Generic;

using Android.App;
using Android.Content;
using Android.Views;
using Android.Widget;
using Android.Graphics;

namespace VorratsUebersicht
{
    public class ArticleListViewAdapter : BaseAdapter<ArticleListView>
    {
        List<ArticleListView> items;
        Activity context;

        public event EventHandler OptionMenu;

        public ArticleListViewAdapter(Activity context, List<ArticleListView> items) : base()
        {
            this.context = context;
            this.items = items;
        }
        public override long GetItemId(int position)
        {
            return position;
        }
        public override ArticleListView this[int position]
        {
            get { return items[position]; }
        }
        public override int Count
        {
            get { return items.Count; }
        }

        public override View GetView(int position, View convertView, ViewGroup parent)
        {
            ArticleListView item = items[position];
            View view = convertView;
            if (view == null) // no view to re-use, create new
                view = context.LayoutInflater.Inflate(Resource.Layout.ArticleListView, null);
            
            view.FindViewById<TextView> (Resource.Id.ArticleListView_Heading).Text    = item.Heading;
            view.FindViewById<TextView> (Resource.Id.ArticleListView_SubHeading).Text = item.SubHeading;
            view.FindViewById<TextView> (Resource.Id.ArticleListView_Notes).Text      = item.Notes;
            view.FindViewById<TextView> (Resource.Id.ArticleListView_Notes).Visibility            = string.IsNullOrEmpty(item.Notes) ? ViewStates.Gone : ViewStates.Visible;
            view.FindViewById<ImageView>(Resource.Id.ArticleListView_OnShoppingList).Visibility   = item.IsOnShoppingList ? ViewStates.Visible : ViewStates.Gone;
            view.FindViewById<TextView> (Resource.Id.ArticleListView_ShoppingQuantity).Visibility = item.IsOnShoppingList ? ViewStates.Visible : ViewStates.Gone;
            view.FindViewById<TextView> (Resource.Id.ArticleListView_ShoppingQuantity).Text       = item.ShoppingQuantity;
            view.FindViewById<ImageView>(Resource.Id.ArticleListView_IsInStorage).Visibility      = item.IsInStorage ? ViewStates.Visible : ViewStates.Invisible;
            view.FindViewById<TextView> (Resource.Id.ArticleListView_StorageQuantity).Visibility  = item.IsInStorage ? ViewStates.Visible : ViewStates.Invisible;
            view.FindViewById<TextView> (Resource.Id.ArticleListView_StorageQuantity).Text        = item.StorageQuantity;

            if ((!item.IsOnShoppingList) && (!item.IsInStorage))
            {
                view.FindViewById<ImageView>(Resource.Id.ArticleListView_IsInStorage).Visibility      = ViewStates.Gone;
                view.FindViewById<TextView> (Resource.Id.ArticleListView_StorageQuantity).Visibility  = ViewStates.Gone;
            }

            TextView option = view.FindViewById<TextView>(Resource.Id.ArticleListView_Option);
            option.Click -= Option_Click;
            option.Click += Option_Click;

            ImageView image = view.FindViewById<ImageView>(Resource.Id.ArticleListView_Image);
            image.Click -= OnImageClicked;
            image.Tag = item.ArticleId;
            image.SetImageResource(Resource.Drawable.ic_photo_camera_black_24dp);
            image.Alpha = 0.25f;

            if (string.IsNullOrEmpty(item.CacheFileName))
                return view;

            var dir = this.context.CacheDir;
            var test = item.CacheFileName;
            
            var cacheFileName = dir.AbsolutePath + "/" + item.CacheFileName + ".png";

            if (File.Exists(cacheFileName))
            {
                Bitmap bitmap = BitmapFactory.DecodeFile(cacheFileName);
                image.SetImageBitmap(bitmap);
                image.Alpha = 2f;
                return view;
            }

            //return view;

            new Thread(new ThreadStart(delegate
            {
                byte[] picture = null;
                try
                {
                    picture = Database.GetArticleImage(item.ArticleId, false)?.ImageSmall;
                }
                catch(Exception)
                {
                    this.context.RunOnUiThread( () =>
                    {
                        image.SetImageResource(Resource.Drawable.baseline_error_outline_black_24);
                        image.Alpha = 0.5f;
                    });
                    return;
                }
                if (picture == null)
                {
                    return;
                }

                File.WriteAllBytes(cacheFileName, picture);

                Bitmap unScaledBitmap = BitmapFactory.DecodeByteArray (picture, 0, picture.Length);
                

                this.context.RunOnUiThread( () =>
                {
                    if (item.ArticleId == (int)image.Tag)
                    {
                        image.SetImageBitmap(unScaledBitmap);
                        image.Alpha = 1f;

                        image.Click += OnImageClicked;
                    }
                });
            })).Start();

            return view;
        }

        private void Option_Click(object sender, EventArgs e)
        {
            this.OptionMenu?.Invoke(sender, EventArgs.Empty);
        }

        private void OnImageClicked(object sender, EventArgs e)
        {
            ImageView imageToView = (ImageView)sender;

            int articleId = (int)imageToView.Tag;

            var articleImage = new Intent(context, typeof(ArticleImageActivity));
            articleImage.PutExtra("ArticleId", articleId);
            context.StartActivity(articleImage);
        }
    }
}

File: app/src/main/java/de/stryi/vorratsuebersicht/ArticleListViewAdapter.kt
package de.stryi.vorratsuebersicht

import android.app.Activity
import android.content.Intent
import android.graphics.BitmapFactory
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.view.menu.MenuBuilder
import androidx.appcompat.widget.PopupMenu
import androidx.recyclerview.widget.RecyclerView
import de.stryi.vorratsuebersicht.database.Records.Article
import de.stryi.vorratsuebersicht.database.Database
import de.stryi.vorratsuebersicht.tools.AddToShoppingListDialog
import kotlinx.coroutines.*

class ArticleListViewAdapter(
    private val articles: List<Article>,
    private val onItemClick: (articleId: Int) -> Unit
) :
    RecyclerView.Adapter<ArticleListViewAdapter.ArticleViewHolder>()
{
    var optionSelect: ((Int, ArticleViewHolder) -> Unit)? = null
    var optionSelected: ((Int, Int) -> Unit)? = null
    private val adapterScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    class ArticleViewHolder(view: View) : RecyclerView.ViewHolder(view)
    {
        val heading: TextView = view.findViewById(R.id.ArticleListView_Heading)
        val subHeading: TextView = view.findViewById(R.id.ArticleListView_SubHeading)
        val notesText:  TextView = view.findViewById(R.id.ArticleListView_Notes)
        val onShoppingList: ImageView = view.findViewById(R.id.ArticleListView_OnShoppingList)
        val shoppingQuantity: TextView = view.findViewById(R.id.ArticleListView_ShoppingQuantity)
        val isInStorage: ImageView = view.findViewById(R.id.ArticleListView_IsInStorage)
        val storageQuantity: TextView = view.findViewById(R.id.ArticleListView_StorageQuantity)
        val image: ImageView = view.findViewById(R.id.ArticleListView_Image)
        val option: TextView = view.findViewById(R.id.ArticleListView_Option)
        val articleListViewOption: TextView = view.findViewById(R.id.ArticleListView_Option)

        var minQuantity: Int?  = null
        var prefQuantity: Int? = null
        var imageJob: Job? = null
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ArticleViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.article_list_view, parent, false)
        return ArticleViewHolder(view)
    }

    override fun onBindViewHolder(holder: ArticleViewHolder, position: Int) {
        val article = articles[position]

        holder.itemView.tag = article.articleId
        holder.minQuantity = article.minQuantity
        holder.prefQuantity = article.prefQuantity

        holder.heading.text    = article.heading
        holder.subHeading.text = article.articleInfo
        holder.notesText.text  = article.notesText
        holder.notesText.visibility = if (article.notesText.isEmpty()) View.GONE else View.VISIBLE

        holder.onShoppingList.visibility   = if (article.isOnShoppingList) View.VISIBLE else View.GONE
        holder.shoppingQuantity.visibility = if (article.isOnShoppingList) View.VISIBLE else View.GONE
        holder.shoppingQuantity.text       = article.shoppingQuantityText

        holder.isInStorage.visibility     = if (article.isInStorage) View.VISIBLE else View.INVISIBLE
        holder.storageQuantity.visibility = if (article.isInStorage) View.VISIBLE else View.GONE
        holder.storageQuantity.text       = article.storageQuantityText

        // Mehr Platz für Notizen
        if (!article.isOnShoppingList && !article.isInStorage)
        {
            holder.onShoppingList.visibility   = View.GONE
            holder.shoppingQuantity.visibility = View.GONE
            holder.isInStorage.visibility      = View.GONE
            holder.storageQuantity.visibility  = View.GONE
        }

        holder.itemView.setOnClickListener {
            onItemClick(article.articleId)
        }

        holder.itemView.setOnLongClickListener {
            showContextMenu(holder)
            true
        }

        holder.articleListViewOption.setOnClickListener {
            showContextMenu(holder)
        }

        // Bild asynchron laden
        holder.imageJob?.cancel()
        holder.image.setImageResource(R.drawable.photo_camera_24px)
        holder.image.alpha = 0.2f
        holder.image.setOnClickListener { }

        holder.imageJob = adapterScope.launch {
            val bitmap = withContext(Dispatchers.IO) {
                val articleImage = Database.getArticleImage(article.articleId, false)
                articleImage?.imageSmall?.let { bytes ->
                    try {
                        BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                    } catch (e: Exception) {
                        null
                    }
                }
            }

            if (isActive) {
                if (bitmap != null) {
                    holder.image.setImageBitmap(bitmap)
                    holder.image.alpha = 1.0f
                    holder.image.setOnClickListener {
                        val intent = Intent(holder.image.context, ArticleImageActivity::class.java)
                        intent.putExtra("ArticleId", article.articleId)
                        holder.image.context.startActivity(intent)
                    }
                } else {
                    holder.image.setImageResource(R.drawable.photo_camera_24px)
                    holder.image.alpha = 0.2f
                }
            }
        }
    }

    override fun onViewDetachedFromWindow(holder: ArticleViewHolder) {
        super.onViewDetachedFromWindow(holder)
        holder.imageJob?.cancel()
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        adapterScope.cancel()
    }

    @android.annotation.SuppressLint("RestrictedApi")
    fun showContextMenu(holder: ArticleViewHolder) {
        val popupMenu = PopupMenu(holder.itemView.context, holder.option)
        popupMenu.menuInflater.inflate(R.menu.article_list_contextmenu, popupMenu.menu)

        if (popupMenu.menu is MenuBuilder) {
            (popupMenu.menu as MenuBuilder).setOptionalIconsVisible(true)
        }

        popupMenu.setOnMenuItemClickListener { menuItem: MenuItem ->
            val articleId = holder.itemView.tag as Int
            when (menuItem.itemId) {
                R.id.ArticleList_ContextMenu_Lagerbestand -> {
                    
                    this.optionSelected?.invoke(articleId, 1)

                    val storageDetails = Intent(holder.itemView.context, StorageItemInventoryActivity::class.java)
                    storageDetails.putExtra("ArticleId", articleId)
                    holder.itemView.context.startActivity(storageDetails)
                    true
                }
                R.id.ArticleList_ContextMenu_AufEinkaufszettel -> {

                    this.optionSelected?.invoke(articleId, 2)

                    AddToShoppingListDialog.showDialog(
                        holder.itemView.context as Activity,
                        articleId,
                        holder.minQuantity,
                        holder.prefQuantity,
                        { refreshListAction(articleId, 2) }
                    )
                    true
                }
                else -> false
            }
        }
        popupMenu.show()
    }

    private fun refreshListAction(articleId: Int, optionId: Int)
    {
        this.optionSelected?.invoke(articleId, optionId)
    }

    override fun getItemCount(): Int = articles.size
}
